[댓글 기능] 러너덕 미션 제출합니다

Deocksoo opened the pull request:
안녕하세요 서머! 리뷰이 러너덕입니다.
이번 페어와 처음부터 과제를 다시 만들었는데,  생각보다  시간이 많이 걸려서 커밋 로그도 엉망이고 코드도 깔끔하지 못하네요..  빠진 테스트도 꽤 있어서 빠른 시일 내로 마무리 못한 부분을 작성해보려고 합니다.
지난 과제 코드까지 커밋을 촘촘하게 찍으면 이번 과제부분이 어느 부분인지 혼동될 것 같아서 앞부분의 커밋은 약간 긴 텀을 두고 작성했습니다. 이번 과제에 해당하는 커밋은 

[9015575] feat : User와 Article 기능 구현

이 부근 이후부터입니다!
그럼 이번  미션도 잘 부탁드립니다 :)

backlo committed:
[게시글 생성/조회/수정/삭제] 포도당 미션 제출합니다. (#27)

* doc(sample.md): update sample.md including feature list

* feat(gradle): add thymeleaf library

* feat: update build.gradle

* feat(HelloWorld): update HelloWorld controller

* feat: update ArticleController including create article

* feat(areticle) : Update article controller including find article list

* feat(Article) : Update Article Controller find article by Id

* feat: Remove duplicate html tags

* feat(article) : Update Article Controller including update to select article

* feat(article) : update article controller including delete function

* feat: update script including remove article

* feat: fix bug css

* fix(article) : update article controller naming

* fix : cleaning code

* feat(gradle) : add Lombok api and change compile to implementation

* fix : remove duplicate css code in article.html, article-edit.html

* fix(article) : Remove the model to which the articleId is mapped

* fix(article) : update article Entity model because of Lombok library

* fix(article) : create article_config.html because of duplicate css

* fix(article) : remove ArticleRepository line and insert 400 error code into delete method

backlo committed:
feat: init project

backlo committed:
feat: init dependencies

backlo committed:
feat: 템플릿 파일 중복 제거

backlo committed:
[게시판 회원 기능] 포도당 미션 제출합니다. (#58)

* doc(sample.md): update sample.md including feature list

* feat(gradle): add thymeleaf library

* feat: update build.gradle

* feat(HelloWorld): update HelloWorld controller

* feat: update ArticleController including create article

* feat(areticle) : Update article controller including find article list

* feat(Article) : Update Article Controller find article by Id

* feat: Remove duplicate html tags

* feat(article) : Update Article Controller including update to select article

* feat(article) : update article controller including delete function

* feat: update script including remove article

* feat: fix bug css

* fix(article) : update article controller naming

* fix : cleaning code

* feat(gradle) : add Lombok api and change compile to implementation

* fix : remove duplicate css code in article.html, article-edit.html

* fix(article) : Remove the model to which the articleId is mapped

* fix(article) : update article Entity model because of Lombok library

* fix(article) : create article_config.html because of duplicate css

* fix(article) : remove ArticleRepository line and insert 400 error code into delete method

* docs(article) : update README.md

* fix(application.properties) : set properties by MySQL database

* fix : Update Properties and add mysql dependencies and hiebernte

* feat(user) : add user class with user controller test

* fix : refactor article class

* feat : create Use Controller and User Entity with JPA

* feat : create User-list class and do requestMapping of user-info

* fix : update User-list page with showing userName and email and clean duplicate html code

* refactor : clean package and remove setter method of User class

* refactor : create member service layer

* feat(user) : create User serivce with checking email repetition and saving User info and deleting user

* feat(user) : validate id, password, and confirm-password in sign-up layout html page

* refactor(signup) : clean code at signup.html page and add email validation

* fix : add articleId in article controller

* fix : redirect when email repetition error occur and show alert at client

* fix(UserControllerTest) : add email repetition check test

* fix() : add login function and login error handling function

* feat : Create Session about of User

* feat() : Create edit user feature

* feat() : create deleting user feature

* refactor() : clean to blog project with convention

* feat(UserServiceTest) : add update test and delete test

* Revert "[게시글 생성/조회/수정/삭제] 포도당 미션 제출합니다. (#27)"

This reverts commit be200464bd7a529ee99eedf01cd4c7873f84fc24.

* fix : fix to ArticleControllerTest with annotation line

* Revert "fix : fix to ArticleControllerTest with annotation line"

This reverts commit 6958fb776aa5c0c8063df14ca1a1677976d3058a.

* fix(UserServiceTest) : reupload test and delete test

* fix : change field inject to constructor injection in controllers

* fix : add annotation in controllers

* fix : fix to SessionAttribute in UserController

* fix : change try-catch to spring validator

* fix : fix to userRepository with Optional return type and fix DataJpaTest to SpringBootTest annotation

* fix : fix userService to mockito test and refactor code

* fix : move to User Object to confirm password

* fix : fix to user controller and refactor code

* refactor : refactor all code

* refactor : refactor gradle

* fix : fix to controller with duplicate Annotation

backlo committed:
feat: article 생성 및 수정 기능

backlo committed:
feat: User 생성, 수정, 삭제, 로그인 기능 구현

backlo committed:
Revert "[게시판 회원 기능] 포도당 미션 제출합니다. (#58)"

This reverts commit b44012bc77f2d0b3a13242fcccb977754a020efb.

backlo committed:
refactor: clean file

backlo committed:
Merge branch 'testbranch2' into testbranch

backlo committed:
feat: 유저 관련 기능 추가 및 구현 완료

backlo committed:
feat : User랑 Article 기능 구현

backlo committed:
feat: error 페이지 추가

backlo committed:
feat: comment entity 생성 및 article 연관성 추가

backlo committed:
feat: CommentDto 생성

backlo committed:
feat: article.html에 comment 템플릿 추가, comment-edit 템플릿 추가

backlo committed:
feat: CommentService, CommentRepository 생성, 조회, 수정 기능 추가

backlo committed:
feat: CommentController에서 comment의 생성, 조회, 수정 기능 추가

backlo committed:
feat: ArticleService, ArticleController에 comment 양방향 매핑 구현

backlo committed:
refactor: BlogControllerAdvice의 코드 중복 제거 및 MisMatch 핸들러 추가

backlo committed:
feat: GuestInterceptor에서 CommentController로 라우팅 하기 전에 로그인 체크

backlo committed:
feat: CommentServiceTest 구현

backlo committed:
fix: Util 테스트 버그 수정

backlo committed:
refactor: 코드 formatting

Deocksoo committed:
delete: clear files to merge with pair

Deocksoo committed:
Merge branch 'backlo' of https://github.com/backlo/jwp-blog into Deocksoo

jihan805 reviewed (changes requested):
안녕하세요 러너덕~!
시간이 많이 않았지만 잘 구현하셨어요 👍 
코멘트 남긴 것 처럼 빠진 테스트 다음번 요청때는 꼭 작성해주세요!
몇가지 피드백 남겼으니 확인해주세요 :)

backlo committed:
feat: Comment 삭제 기능 구현

backlo committed:
style: 파일 포매팅, 불필요한 log 삭제

backlo committed:
refactor: 자신의 comment인지 체크하는 로직을 comment 엔티티로 이동

backlo committed:
feat: Article, Comment 에 timestamp 추가

backlo committed:
refactor: article의 작성자를 체크하는 로직을 article entity에서 담당하도록 수정

backlo committed:
fix: commentDto 트랜잭션 처리 및 article controller 에 있는 예외 도메인으로 이동

Deocksoo committed:
Merge branch 'backlo' of https://github.com/backlo/jwp-blog into Deocksoo3

Deocksoo committed:
feat: 프로젝트에 ide 롬복 설정을 추가합니다

Deocksoo committed:
fix: application.properties의 db설정을 로컬 환경에 맞게 변경합니다

Deocksoo committed:
refactor: 댓글 수정, 삭제 버튼의 모양과 위치를 재설정합니다

Deocksoo committed:
refactor: 댓글이 작성된 순서대로 출력될 수 있게 변경합니다

Deocksoo committed:
feat: BolgControllerAdvice에서 게시글과 댓글을 찾지 못하는 예외를 처리해줄 수 있도록 수정합니다

Deocksoo committed:
refactor: CommentService의 save 메서드에서 과다한 메서드 체이닝을 제거합니다

Deocksoo committed:
refactor: Comment에서 의존성을 갖고 있는 Article을 Lazy Loading하도록 수정합니다

Article과 comment가 의존성을 갖고 있으므로 EAGER 방식으로 fetch하게 되면
index 페이지를 로딩할 때 join 쿼리가 날아가게 됩니다. 이 때 article과 comment의 수가 많아질수록
join 쿼리를 날릴 때 비용이 커지는 데다가, 사용하지 않는 Comment의 데이터까지 모두 받아오게 되므로
효율적이지 못합니다. 이러한 이유에서 LAZY 로딩으로 변경하였습니다.

Deocksoo committed:
refactor: Article 객체에서 comment List를 조회할 때 unmodifiable 객체를 리턴하도록 수정합니다

Deocksoo committed:
test: Article 도메인 객체 테스트를 추가합니다

Deocksoo committed:
refactor: Article 객체 외부에서 항상 unmodifiable comment list를 사용하도록 수정합니다

Deocksoo committed:
refactor: Comment 객체의 updateArticle 메서드를 제거합니다

Article로 update하는 메서드를 따로 만들지 않고 앞서 만들어 둔 update를 통해 Article도
update하도록 수정하였습니다.

Deocksoo committed:
test: UserTest와 CommentTest를 추가합니다

Deocksoo committed:
refactor: 회원가입시 email의 중복 체크에 @column unique 속성을 활용하도록 수정합니다

Deocksoo committed:
refactor: 불필요한 import를 제거합니다

Deocksoo committed:
test: BlogBodyContentsSpec을 이용해 테스트 코드의 response body 컨텐츠 체크 로직의 코드 중복을 제거합니다

1. BlogBodyContentsSpec 클래스를 선언해 BodyContentSpec을 감싸고, assertThatBodyOf
메서드를 이용해 responseSpec으로부터 BlogBodyContentsSpec을 만들도록 합니다.
2. contains 메서드를 이용해 bodyContentSpec이 임의의 content를 포함하는지 확인할 수
있도록 합니다.
3. 위 두가지 메서드를 이용해 테스트 코드에서 반복되던 response body의 컨텐츠를 체크하던
코드의 중복을 제거하였습니다.

Deocksoo committed:
test: Utils 클래스의 메서드 순서를 변경합니다

Deocksoo committed:
refactor: ArticleController의 findArticle 메서드명을 fetchArticle로 변경합니다

Deocksoo committed:
refactor: UserService delete 메서드 argument 이름을 변경합니다

Deocksoo committed:
refactor: ArticleTest의 addComment 메서드 변수명 변경

Deocksoo committed:
refactor: UserController delete 메서드 순서를 변경합니다

Deocksoo committed:
refactor: ServiceTest들의 static 상수 중복을 제거합니다

Deocksoo committed:
refactor : GuestInterceptor isSave 메서드명을 isSignUp으로 변경합니다

Deocksoo commented:
안녕하세요 서머, 좋은 피드백 감사드려요 :)
아래에 피드백  반영한 내용을 간단하게 정리해보았어요! 

1. 말씀해주신 fetchType => Lazy에 대해 조금 찾아보았습니다! 현재에는 코멘트를 처음 조회할 때(=특정 게시글을 클릭할 떄)  작성자와 항상 같이 표시되고 있어서  Eager loading 하도록 하였습니다. 하지만 Article과 comment는 Eager loading 하면 처음 index page를 가져올 때 모든 게시글과 함께 모든 comment까지 조회하게 되어 게시글과 comment가 많아질수록 성능에 문제가 생길것 같다고 생각했습니다. 그래서 이 부분만 LAZY로 수정해 보았어요!

2. 그 외에도 언급해주셨던 피드백을 모두 반영해보는 것이 좋을 것 같아 수정했습니다.
 
3. 이번에 피드백해주신 부분을 수정하면서 따로 조금 더 고민해보고 싶었던 부분은 Test 코드의 중복을 제거하는 것이었어요. WebTestClient를 사용하면서 response body의 contents를 확인할 때 중복되어 계속 등장하는 코드들이 있는데, 이 부분의 중복이 계속 마음에 걸렸어요. 그래서 BlogBodyContentSpec을 테스트 utils로 만들어 해당 중복을 제거해보려고 했는데, 이렇게 만드는 것이 테스트코드를 읽기 쉽게 하는 데에 효과적인지 아직 잘 모르겠습니다. 혹시 의견 부탁드려도 될까요?

Deocksoo committed:
feat: gradlew build를 위한 setting들을 추가합니다

jihan805 reviewed (approved):
안녕하세요 러너덕 ~
피드백 반영 잘해주셨어요 :)
간단한 코멘트 남겼으니 확인해주세요, 머지할게요 👍

Deocksoo commented:
헉 ㅠㅠ 늦은 시간까지 너무 감사합니다. 말씀해주신 부분들 추가 반영 하도록 하겠습니다!
감사합니다 :)
